### PR TITLE
Add SVMLight DataSet

### PR DESCRIPTION
**Pull request by vilozio**
## Description
Add DataSet for svmlight/libsvm files using scikit-learn library as backend.

Resolves kedro-org/kedro#1972

## Development notes
Added a new data set for libsvm sparse data format. Tested it on my files and added unit tests for coverage.

## Checklist

- [x] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [x] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [x] Added tests to cover my changes

<a href="https://gitpod.io/#https://github.com/kedro-org/kedro/pull/1992"></a>

### CODE REVIEW

File: test_requirements.txt [hunk 1]
```diff
@@ -49,6 +49,8 @@ redis~=4.1
 requests-mock~=1.6
 requests~=2.20
 s3fs>=0.3.0, <0.5  # Needs to be at least 0.3.0 to make use of `cachable` attribute on S3FileSystem.
+scikit-learn~=1.1
+scipy~=1.9
```
Comment (merelcht) [inline test_requirements.txt:53]: You'll also need to add this dataset to [`setup.py`](https://github.com/kedro-org/kedro/blob/main/setup.py) together with those requirements.

File: kedro/extras/datasets/svmlight/svmlight_dataset.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+)
+
+_DI = Tuple[Union[ndarray, csr_matrix], ndarray]
+"""Type of the data input."""
+_DO = Tuple[csr_matrix, ndarray]
+"""Type of the data output."""
+
+
+class SVMLightDataSet(AbstractVersionedDataSet[_DI, _DO]):
+    """``SVMLightDataSet`` loads/saves data from/to a svmlight/libsvm file using an
```
Comment (merelcht) [inline kedro/extras/datasets/svmlight/svmlight_dataset.py:29]: So if I understand it correctly, this dataset can be used to load and save data in both the `svmlight` and `libsvm` formats? Is there a difference in how users should use the dataset depending on the format they want to use?

**Review by merelcht — COMMENTED** [review]
Thank you for this contribution @vilozio ! ⭐  I've left a couple of comments/questions, but in general this implementation looks good 🙂

File: kedro/extras/datasets/svmlight/svmlight_dataset.py (hunk 2 above)
Comment (AhdraMeraliQB) [reply kedro/extras/datasets/svmlight/svmlight_dataset.py:33]: From what I've been able to find, there is no difference, both formats will be loaded and saved to `.svm` files through the [sklearn implementation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_svmlight_file.html).

**Review by AhdraMeraliQB — APPROVED** [review]
Hi @vilozio, thank you so much for your contribution! Just took over to fix a few things and get it past the finish line - awesome work! 🌟

File: kedro/extras/datasets/svmlight/svmlight_dataset.py (hunk 2 above)
Comment (AhdraMeraliQB) [reply kedro/extras/datasets/svmlight/svmlight_dataset.py:33]: @merelcht From what I've been able to find, there is no difference, both formats will be loaded and saved to `.svm` files through the [sklearn implementation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_svmlight_file.html).

File: kedro/extras/datasets/svmlight/svmlight_dataset.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    DataSetError,
+    Version,
+    get_filepath_str,
+    get_protocol_and_path,
+)
+
+_DI = Tuple[Union[ndarray, csr_matrix], ndarray]
+"""Type of the data input."""
+_DO = Tuple[csr_matrix, ndarray]
+"""Type of the data output."""
```
[the remark this answers is not in the document]
Comment (AhdraMeraliQB) [reply kedro/extras/datasets/svmlight/svmlight_dataset.py]: Done! [4f88b4f](https://github.com/kedro-org/kedro/pull/1992/commits/4f88b4f569c08afeea0a7c0fd370df8a0130b634)

File: kedro/extras/datasets/svmlight/svmlight_dataset.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    Example using Python API:
+    ::
+
+        >>> from kedro.extras.datasets.svmlight import SVMLightDataSet
+        >>> import numpy as np
+        >>>
+        >>> # Features and labels.
+        >>> data = (np.array([[0, 1], [2, 3.14159]]), np.array([7, 3]))
+        >>>
+        >>> # data_set = SVMLightDataSet(filepath="gcs://bucket/test.svm")
```
[the remark this answers is not in the document]
Comment (AhdraMeraliQB) [reply kedro/extras/datasets/svmlight/svmlight_dataset.py]: Fixed - removed the line in [4f88b4f](https://github.com/kedro-org/kedro/pull/1992/commits/4f88b4f569c08afeea0a7c0fd370df8a0130b634)

File: kedro/extras/datasets/svmlight/svmlight_dataset.py (hunk 2 above)
Comment (vilozio) [reply kedro/extras/datasets/svmlight/svmlight_dataset.py:33]: As far as I know the `svmlight` format is an extension of `libsvm` fromat.

LIBSVM library is used for SVM classification and regression. Original `libsvm` format looks like this

```
<label> <index1>:<value1> <index2>:<value2> ...
```

In addition, SVM-Light library solves ranking tasks. Ranking problems require inputs to be grouped into query groups. So `svmlight` format also added an optional parameter for an input  - query id (qid). Inputs with the same qid belong to the same group.

```
3 qid:1 1:0.53 2:0.12
2 qid:1 1:0.13 2:0.1
7 qid:2 1:0.87 2:0.12
```

This is the only difference between them that I know.

The names for these formats are given after the libraries where they are coming from, but they are still used with other Gradient Boosting Machines libraries. For example, XGBoost extended further the format and added an [optional weight parameter to labels](https://xgboost.readthedocs.io/en/stable/tutorials/input_format.html#instance-weights)

```
1:1.0 101:1.2 102:0.03
0:0.5 1:2.1 10001:300 10002:400
```

This extension is not supported by sklearn.

Comment (merelcht) [reply kedro/extras/datasets/svmlight/svmlight_dataset.py:33]: Thanks for explaining this in more detail @vilozio 🙂

File: kedro/extras/datasets/svmlight/svmlight_dataset.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+class SVMLightDataSet(AbstractVersionedDataSet[_DI, _DO]):
+    """``SVMLightDataSet`` loads/saves data from/to a svmlight/libsvm file using an
+    underlying filesystem (e.g.: local, S3, GCS). It uses sklearn functions
+    ``dump_svmlight_file`` to save and ``load_svmlight_file`` to load a file.
+
+    Data is loaded as a tuple of features and labels. Labels is NumPy array,
+    and features is Compressed Sparse Row matrix.
+
+    This format is a text-based format, with one sample per line. It does
+    not store zero valued features hence is suitable for sparse dataset.
```
Comment (merelcht) [inline kedro/extras/datasets/svmlight/svmlight_dataset.py:41]:
```suggestion
    not store zero valued features hence it is suitable for sparse datasets.
```

**Review by merelcht — APPROVED** [review]
Left one more minor comment, but other than that this looks good to go! Thanks so much for this contribution @vilozio and @AhdraMeraliQB for finishing it up ⭐ ⭐

File: kedro/extras/datasets/svmlight/svmlight_dataset.py (hunk 5 above)
Comment (vilozio) [reply kedro/extras/datasets/svmlight/svmlight_dataset.py]: Fixed in d8ebc47c0827adde97bb94c70f5b7906fdd6a54b
